### PR TITLE
Updates SAML attribute definitions to be used by dictionary

### PR DESCRIPTION
**Pull request by arothuis**
Part of (preparations for) [ticket 106955902] (https://www.pivotaltracker.com/story/show/106955902)

Done:
* Add known URNs
* Missing URNs are set to null.
* Add logic to allow null URNs
* Move logic to responsible classes

### CODE REVIEW

**Comment by arothuis** [comment]
@rjkip, @DRvanR; see PR description/comment, thoughts?

**Comment by DRvanR** [comment]
Before merging these issues have to be resolved with SURF. 

**Missing urn**
I'd say add the ability that an attribute can be defined based on either a urn:mace, a urn:oid, or both (in that case the value should by `null` rather than `''`).

**Duplicated URNs for different attributes**
My suggestion would be to add them all as different attribute definitions. There are two ways to do this:

1.  add each attribute with a `null` value for the duplicated urn, possibly with one attribute actually having that urn (preference based)
2.  add each attribute with the same duplicated urn, and expand the lookup logic to check if there are multiple attribute definitions with the same urn (based on which the attribute definition is searched), and if so return the preferred value (additional argument), logging that multiple definitions have been found.
3.  add each attribute with the same duplicated urn, and expand the lookup logic to check if there are multiple attribute definitions with the same urn (based on which the attribute definition is searched), and if so trigger an error (Exception). 

The 3rd option has my preference, the 1st option is the simplest but also just ignores the problem at hand. I would recommend against the 2nd option due to the amount of work involved vs result.

Do you have any ideas how the current implementation handles these cases?

**Comment by arothuis** [comment]
Currently, attribute definitions are added as follows: a JSON-file is loaded, decoded and handed over to a definition [denormalizer](https://github.com/OpenConext/OpenConext-engineblock/blob/master/library/EngineBlock/Attributes/Definition/Denormalizer.php). There, aliases that don't have a definition are resolved and get the definition of the original attribute name. This results in a config array with each possible URN as a key and constraints/descriptions/comments as `array` values or a `string` reference to another URN that should contain said `array` values.

Attributes, when parsed, get normalized via an attribute [normalizer](https://github.com/OpenConext/OpenConext-engineblock/blob/library/EngineBlock/Attributes/Normalizer.php). There, references are followed as well (not sure why).

**Missing URN**
A missing urn will not be treated: it is not present. 

**Duplicated URNs for different attributes**
Duplications (aliases) get their own definition in the config array from the attribute name that has a definition upon application boot.

Our implementation differs, since we are returning names and their aliases.

**Comment by DRvanR** [comment]
As discussed, the _missing urns_ will be resolved in the proposed way, the _duplicate urns_ are actually not duplicated urns and therefore do not require any action

**Comment by arothuis** [comment]
@rjkip :fork_and_knife:

File: src/SAML2/Attribute/Attribute.php [hunk 1]
```diff
@@ -28,16 +30,27 @@ class Attribute
...
+     * @param string[] $value
      */
-    public function __construct(AttributeDefinition $attributeDefinition, array $values)
+    public function __construct(AttributeDefinition $attributeDefinition, array $value)
     {
+        if ($attributeDefinition->getMultiplicity() === AttributeDefinition::MULTIPLICITY_SINGLE
+            && count($value) > 1
+        ) {
+            throw new UnexpectedValueException(sprintf(
+                'AttributeDefinition "%s" has a single-value multiplicity, yet returned'
```
Comment (rjkip) [inline src/SAML2/Attribute/Attribute.php]: > returned

File: src/Tests/SAML2/Response/Assertion/AssertionAdapterTest.php [hunk 2]
```diff
@@ -103,16 +105,14 @@ public function no_presence_of_attribute_can_be_confirmed_if_no_attribute_defini
...
 
+        // empty dictionary
+        $dictionary = new AttributeDictionary();
         $adapter      = new AssertionAdapter($assertion, $dictionary);
         $attributeSet = $adapter->getAttributeSet();
 
-        $attributeIsNotInSet = !$attributeSet->contains($attributeExpectedNotToBeContained);
-
-        $this->assertTrue($attributeIsNotInSet, 'Expected attribute to not be part of AttributeSet, but it is');
+        $attributeSet->contains($attributeExpectedNotToBeContained);
```
Comment (rjkip) [inline src/Tests/SAML2/Response/Assertion/AssertionAdapterTest.php]: Should there be an assertion here?

Comment (arothuis) [inline src/Tests/SAML2/Response/Assertion/AssertionAdapterTest.php]: `@expectedException`

Comment (rjkip) [inline src/Tests/SAML2/Response/Assertion/AssertionAdapterTest.php]: :+1:

**Comment by rjkip** [comment]
:sun_with_face: :ship:

File: src/SAML2/Attribute/AttributeDefinition.php [hunk 3]
```diff
@@ -51,18 +51,18 @@ class AttributeDefinition
...
-            throw InvalidArgumentException::invalidType('string', 'urnMace', $urnMace);
+        if (!is_null($urnMace) && !is_string($urnMace)) {
+            throw InvalidArgumentException::invalidType('null or string', 'urnMace', $urnMace);
         }
 
-        if (!is_string($urnOid)) {
-            throw InvalidArgumentException::invalidType('string', 'urnOid', $urnOid);
+        if (!is_null($urnOid) && !is_string($urnOid)) {
+            throw InvalidArgumentException::invalidType('null or string', 'urnOid', $urnOid);
         }
```
Comment (DRvanR) [inline src/SAML2/Attribute/AttributeDefinition.php]: seems we've forgot the check that at least one of `mace` and `oid` must be set.

Comment (arothuis) [inline src/SAML2/Attribute/AttributeDefinition.php]: Good catch, added an exception

**Comment by DRvanR** [comment]
cheers, :+1:

**Pull request merged**